### PR TITLE
Update decomp for upsample_bilinear2d_vec

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #87176

This new implementation fuses multiple index operations into one.

### CODE REVIEW

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -1902,13 +1902,21 @@ def upsample_bilinear2d_vec(
...
-    v3 = input[:, :, x_floor_view, y_ceil]
-    v4 = input[:, :, x_ceil_view, y_ceil]
+    x_floor_view = x_floor.view(1, -1, 1)
+    x_ceil_view = x_ceil.view(1, -1, 1)
+    y_floor_view = y_floor.view(1, 1, -1)
+    y_ceil_view = y_ceil.view(1, 1, -1)
+
+    x_indices = torch.cat([x_floor_view, x_ceil_view, x_floor_view, x_ceil_view], dim=0)
+    y_indices = torch.cat([y_floor_view, y_floor_view, y_ceil_view, y_ceil_view], dim=0)
+    v_indices = torch.arange(4, dtype=torch.int64, device=input.device).view(4, 1, 1)
```
Comment (lezcano) [inline torch/_decomp/decompositions.py:1912]: This was a tad difficult to read. A better way to write this one would be:
```suggestion
    x_floor_view = x_floor.unsqueeze(1)
    x_ceil_view = x_ceil.unsqueeze(1)
    y_floor_view = y_floor.unsqueeze(0)
    y_ceil_view = y_ceil.unsqueeze(0)

    x_indices = torch.stack([x_floor_view, x_ceil_view, x_floor_view, x_ceil_view], dim=0)
    y_indices = torch.stack([y_floor_view, y_floor_view, y_ceil_view, y_ceil_view], dim=0)
    v_indices = torch.arange(4, dtype=torch.int64, device=input.device)
    v_indices = _unsqueeze_to_dim(v_indices, 3)
```

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -1902,13 +1902,21 @@ def upsample_bilinear2d_vec(
...
+
+    input = input.unsqueeze(2).expand(n_batch, n_channels, 4, in_h, in_w)
+    v = input[:, :, v_indices, x_indices, y_indices]
+    v1 = v[:, :, 0, :, :]
+    v2 = v[:, :, 1, :, :]
+    v3 = v[:, :, 2, :, :]
+    v4 = v[:, :, 3, :, :]
```
Comment (lezcano) [inline torch/_decomp/decompositions.py:1919]: You can simply do
```suggestion
    v1, v2, v3, v4 = torch.unbind(v, dim=2)
```
We should have an alias called `unstack` for this function. I never remember its name.

**Review by lezcano — COMMENTED** [review]
I think it was @fdrocha that implemented this one, so I'll let him review it, but at first sight this Is `torch.cat` now efficient when the args have the same shape? If this is the case, this decomp should now generate an efficient backwards as well, which is nice. Could you post the generated triton code for the forward and the backward and the running times?

cc @Chillee.

**Comment by fdrocha** [comment]
> I think it was @fdrocha that implemented this one, so I'll let him review it, but at first sight this Is `torch.cat` now efficient when the args have the same shape? If this is the case, this decomp should now generate an efficient backwards as well, which is nice. Could you post the generated triton code for the forward and the backward and the running times?
> 
> cc @Chillee.

This wasn't me, it seems it was @vivekkhandelwal1

**Comment by ngimel** [comment]
How is the perf of this one with inductor compared to current version? And the generated backward?

**Comment by SherlockNoMad** [comment]
Here's generated triton kernel for forward: https://gist.github.com/SherlockNoMad/61223e5bb994303c42509e66a25b242e

It results in 4 kernels, which is worse than the original impl. 
In addition, it error out with "Triton Error [CUDA]: invalid argument" when I run it.

Need more investigation...

**Comment by ngimel** [comment]
Hm, I can run it, but yeah 4 kernels is bad.

**Comment by ngimel** [comment]
The problem I think is `stack` that inductor doesn't handle very well

**Comment by Chillee** [comment]
@ngimel If we had that `concat` optimization I think we'd able to use stack here :P

**Comment by lezcano** [comment]
Now we have the concat optimisation. Perhaps this PR is worth revisiting? @SherlockNoMad

**Comment by ngimel** [comment]
Wait, where is the concat optimization?

**Comment by lezcano** [comment]
I thought the concat optimisation was the one performed by, say, this code:
```python
import torch
from torch._dynamo import optimize


@optimize("inductor")
def sincos(x):
    return torch.cat((x.sin(), x.cos()), dim=-1)


x = torch.randn(64, 64, device="cuda")
sincos(x)
```

**Comment by ngimel** [comment]
It was present for a long time

**Pull request closed without merging**